src/Clash/Cores/Etherbone/Base.hs

```diff
@@ -0,0 +1,291 @@
+  -- This can be used to wait for all operations to be finished, including
+  -- writes, before sending the response. The default behaviour for now is to
+  -- not wait for all writes to finish.
+  { _resDat  :: Maybe dat
+  -- | Forwarded End-of-record flag
+  , _resEOR  :: Bool
+  -- | Forwarded End-of-packet flag
+  , _resEOP  :: Bool
+  } deriving (Generic, Show, NFDataX, ShowX, NFData, Eq)
+
+-- | The data sent over the bypass line.
+-- It is not assured that the data on the bypass line is in sync with the data
+-- returned from the WishboneMaster. This means that the RecordBuilder should
+-- save the header and base address in its state.
+--
+-- By using a bypass line the state machine of the WishboneMaster can be made
+-- simpler and the RecordBuilder can know about the number of reads and writes
+-- before an operation is finished.
+data Bypass addrWidth = Bypass
+  -- | The 'RecordHeader' from the @_meta@ field.
+  { _bpHeader :: RecordHeader
+  -- | The @BaseRetAddr@ that needs to be returned if a read operation has is
+  -- sent.
+  , _bpBase   :: Maybe (BitVector addrWidth)
+  -- | The @_abort@ signal from the PacketStream. If this is @True@, no new data
+  -- will come from 'wishboneMasterC' and 'recordBuilderC' should be set back
+  -- into its initial state.
+  , _bpAbort  :: Bool
+  } deriving (Generic, NFDataX, NFData, Show, ShowX, Eq)
+
+type ByteSize dat = DivRU (BitSize dat) 8
+
+-- | Extract 'EBHeader' data from a @PacketStream@ into the metadata.
+-- With a 64-bit bus this shifts the data-stream by 4 bytes.
+etherboneDepacketizerC :: forall dom dataWidth .
+ ( HiddenClockResetEnable dom, KnownNat dataWidth, 4 <= dataWidth )
+ => Circuit (PacketStream dom dataWidth ()) (PacketStream dom dataWidth EBHeader)
+etherboneDepacketizerC = depacketizerC @_ metaMap
+  where
+    metaMap :: EBHeader -> () -> EBHeader
+    metaMap hdr _ = hdr
+
+-- | Concat 'EBHeader' metadata back into the data stream. This gives
+-- backpressure for one cycle.
+etherbonePacketizerC :: forall dom dataWidth .
+  ( HiddenClockResetEnable dom, KnownNat dataWidth, 4 <= dataWidth )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth ())
+etherbonePacketizerC = packetizerC (const ()) metaMap
+  where
+    metaMap :: EBHeader -> EBHeader
+    metaMap hdr = hdr
+
+-- | Strips 4 bytes from the start of a packet in the case of a 64-bit wishbone
+-- bus and passes the input for a 32-bit bus.
+-- This is needed to get rid of the 'potential padding' in the case of a 64-bit
+-- bus.
+etherbonePaddingStripperC :: forall dom dataWidth .
+  ( HiddenClockResetEnable dom
+  , KnownNat dataWidth
+  , 4 <= dataWidth
+  )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth EBHeader)
+etherbonePaddingStripperC = case compareSNat (SNat @dataWidth) d4 of
+  SNatLE -> idC
+  SNatGT -> depacketizerC metaMap
+  where
+    metaMap :: Vec (dataWidth-4) (BitVector 8) -> EBHeader -> EBHeader
+    metaMap _ hdr = hdr
```

Comment:
The alternative would be (or with `flip const`)
```
SNatGT -> depacketizerC @_ @(Vec (dataWidth-4) (BitVector 8)) (\_ x -> x)
```
I am not sure if that is more readable.

Reply:
I'd say in this specific instance you can do
```suggestion
  SNatGT -> depacketizerC @(dataWidth-4) (flip const)
```
because the first tyarg, `headerBytes`, is necessarily equal to the length of the vector. But it can't be done without specifying a type somewhere. In such a case, it might be better to write it out just as you did here, with a function `metaMap` and its type. Even though your function is equal to a specific specialisation of `flip const`, readability often trumps other concerns.

Reply:
`depacketizerC` (and `packetizerC`) need to know the bitsize of the header and the stripped data as well. With this suggestion it is missing `BitPack` and the `KnownNat` of `BitSize` of the header. 

With `depacketizerC @_ @(Vec (dataWidth-4) _)`  we still miss the bitsize contraint (`typecheck: • Could not deduce (((dataWidth - 4) * BitSize w1) ~ (w0 * 8))`).

While these can probably be fixed by adding more constraints to `etherbonePaddingStripperC` and using type application to set `header` in `depacketizerC`, I think I prefer what I had originally as that is much more readable to me.

Reply:
You're absolutely right, I made a silly mistake, sorry. I also prefer your original solution.